snuba/subscriptions/consumer.py

```diff
@@ -0,0 +1,387 @@
+            return self.__state
+        else:
+            raise ValueError("unexpected state")
+
+
+class TaskSetConsumer(Consumer[TaskSet]):
+    """
+    A ``Consumer`` class that implements an API that is similar to the
+    ``Message``-based API, except that it returns a ``TaskSet`` instance
+    rather than a Kafka ``Message``.
+
+    This consumer uses the progression of time in a Kafka partition as a
+    "virtual clock" rather than depending on wall clock time. This allows
+    replaying history without gaps in task execution if the consumer
+    is not running when a task would have been scheduled to execute for
+    increased reliability (no schedule gaps) and easier testing on historical
+    data.
+
+    This consumer differs from a typical consumer in the way that messages
+    are processed: the tasks that are contained within a task set are
+    identified by whether or not they fall within a time interval formed by
+    the timestamps of the last two messages received within a partition.
+
+    The first message received will not cause a `TaskSet` to be returned by
+    the consumer, since only one endpoint of the interval can be established.
+
+    The second message received by the consumer will cause a `TaskSet` to be
+    created with all tasks scheduled between the time interval represented by
+    the timestamps of the previous two messages consumed within that
+    partition. (This interval's lower bound is exclusive, and the upper bound
+    is inclusive.) All subsequent messages recieved by the consumer will
+    similarly cause a ``TaskSet`` to be created with all tasks scheduled
+    between the timestamps of the previous two messages consumed within that
+    partition as well.
+
+    This consumer requires that the ``message.timestamp.type`` topic
+    configuration is set to ``LogAppendTime`` for the topics that it is
+    subscribed to, so that that the message time is set by the primary broker
+    for the topic -- not the producer of the message -- ensuring that each
+    partition timestamp moves monotonically.
+
+    There is also a slight difference between the vanilla ``Consumer``
+    behavior and this implementation during rebalancing: whenever a partition
+    is assigned to this consumer, offsets are *always* automatically reset to
+    the committed offset for that partition (or if no offsets have been
+    committed for that partition, the offset is reset in accordance with the
+    ``auto.offset.reset`` configuration value.) This causes partitions that
+    are maintained across a rebalance to have the same offset management
+    behavior as a partition that is moved from one consumer to another. To
+    prevent uncommitted messages from being consumed multiple times,
+    ``commit`` should be called in the partition revocation callback.
+
+    This consumer can either be used with automatic timed offset commit (the
+    librdkafka default, set using the ``enable.auto.commit`` configuration
```

Comment:
I assume this means that enable.auto.commit has, upon restart, the same behavior than if we consumer manually:
consume offset 5
commit offset 5
restart
consumer offset 5

Is this correct?

Reply:
Yes. Is this worth calling out specifically?